usl_models/usl_models/flood_ml/dataset.py

```diff
@@ -0,0 +1,231 @@
+import logging
+import random
+from typing import Iterator, Optional, Tuple
+import urllib.parse
+
+from google.cloud import firestore
+from google.cloud import storage
+import numpy
+from numpy.typing import NDArray
+import tensorflow as tf
+
+from usl_models.flood_ml import constants
+from usl_models.flood_ml import metastore
+from usl_models.flood_ml import model
+
+
+def load_dataset_windowed(
+    sim_names: list[str],
+    batch_size: int = 4,
+    n_flood_maps: int = constants.N_FLOOD_MAPS,
+    m_rainfall: int = constants.M_RAINFALL,
+    max_chunks: Optional[int] = None,
+    firestore_client: Optional[firestore.Client] = None,
+    storage_client: Optional[storage.Client] = None,
+) -> tf.data.Dataset:
+    """Creates a dataset which generates chunks for flood model training.
+
+    The examples are generated from multiple simulations.
+    They are windowed for training on next-map prediction.
+    The dataset iteratively yields examples read from Google Cloud Storage to avoid
+    pulling all examples into memory at once.
+
+    Args:
+      sim_names: The simulation labels to use for training,
+                 e.g. ["Manhattan-config_v1/Rainfall_Data_1.txt"]
+      batch_size: Size of batches yielded by the dataset.
```

Comment:
Let's leave some memory hints here or point it at a doc that has more info on how to select batch size. E.g. "approx VRAM requirement = 10GB * batch_size during training".